### PR TITLE
[SCB-327] Update metrics publish data module (Re-organized)

### PR DESCRIPTION
**Pull request by zhengyangyong**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
**Re-organize commits of PR : https://github.com/apache/incubator-servicecomb-java-chassis/pull/550**

Adjust metrics publish format to spring cloud netflix style
Before:
servicecomb.instance.system.cpu.load
servicecomb.instance.system.cpu.runningThreads
servicecomb.instance.system.heap.init
servicecomb.instance.system.heap.max
servicecomb.instance.system.heap.commit
servicecomb.instance.system.heap.used
servicecomb.instance.system.nonHeap.init
servicecomb.instance.system.nonHeap.max
servicecomb.instance.system.nonHeap.commit
servicecomb.instance.system.nonHeap.used
servicecomb.instance &#124; operationName.producer.waitInQueue.count
servicecomb.instance &#124; operationName.producer.lifeTimeInQueue.average
servicecomb.instance &#124; operationName.producer.lifeTimeInQueue.max
servicecomb.instance &#124; operationName.producer.lifeTimeInQueue.min
servicecomb.instance &#124; operationName.producer.executionTime.average
servicecomb.instance &#124; operationName.producer.executionTime.max
servicecomb.instance &#124; operationName.producer.executionTime.min
servicecomb.instance &#124; operationName.producer.producerLatency.average
servicecomb.instance &#124; operationName.producer.producerLatency.max
servicecomb.instance &#124; operationName.producer.producerLatency.min
servicecomb.instance &#124; operationName.producer.producerCall.total
servicecomb.instance &#124; operationName.producer.producerCall.tps
servicecomb.instance &#124; operationName.consumer.consumerLatency.average
servicecomb.instance &#124; operationName.consumer.consumerLatency.max
servicecomb.instance &#124; operationName.consumer.consumerLatency.min
servicecomb.instance &#124; operationName.consumer.consumerCall.total
servicecomb.instance &#124; operationName.consumer.consumerCall.tps

Current:
jvm(statistic=gauge,name={name})
{name} include :
**cpuLoad,cpuRunningThreads,heapInit,heapMax,heapCommit,
heapUsed,nonHeapInit,nonHeapMax,nonHeapCommit,nonHeapUsed**

servicecomb.invocation(operation={operationName},role={role},stage={stage},statistic={statistic},status={status},unit={unit})

| TagName       | TagValue                  | 
| :---------- | :--------------------- |
| operation | Operation Name |
| role | consume,producer |
| stage | queue,execution,total |
| statistic | tps,count,max,waitInQueue,latency |
| status | statusCode like 200, 404 ... etc|
| unit| time unit like MILLISECONDS etc |

### CODE REVIEW

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java [hunk 1]
```diff
@@ -188,15 +188,16 @@ public String getMicroserviceQualifiedName() {
...
 
-  public void triggerFinishedEvent(int statusCode, boolean success) {
+  public void triggerFinishedEvent(int statusCode) {
     long finishedTime = System.nanoTime();
     EventUtils
-        .triggerEvent(new InvocationFinishedEvent(operationMeta.getMicroserviceQualifiedName(),
-            this.invocationType, finishedTime - startProcessingTime,
-            finishedTime - startTime, statusCode, success));
+        .triggerEvent(new InvocationFinishedEvent(operationMeta.getMicroserviceQualifiedName(), this.invocationType,
+            startProcessingTime - startTime,
```
Comment (WillemJiang) [inline core/src/main/java/org/apache/servicecomb/core/Invocation.java]: Can you explain the meaning of startTime, startProcessingTime?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/DefaultHealthCheckExtraData.java [hunk 2]
```diff
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.servicecomb.metrics.common;
+package org.apache.servicecomb.foundation.metrics.publish;
 
 public class DefaultHealthCheckExtraData {
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/DefaultHealthCheckExtraData.java]: What's the relationship between the DefaultHealthCheckExtraData and HealthCheckResult.extraData?
We don't need to use the String to store result,  it's easy to use toString method to turn the instance of HealthCheckExtraData into a string.

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricNode.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.servicecomb.foundation.metrics.MetricsConst;
+
+public class MetricNode {
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricNode.java]: Can you show me the unit test of MetricNode?

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricsLoader.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      Metric metric = new Metric(entry.getKey(), entry.getValue());
+      this.metrics.computeIfAbsent(metric.getName(), m -> new ArrayList<>()).add(metric);
+    }
+  }
+
+  public MetricNode getMetricTree(String id, String... groupTagKeys) {
+    if (metrics.containsKey(id)) {
+      return new MetricNode(metrics.get(id), groupTagKeys);
+    }
+    return null;
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricsLoader.java]: I'm not a big fan of null, you can throw exception or using Option return the value.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java [hunk 5]
```diff
@@ -43,7 +43,6 @@ public void process(Event data) {
     if (InvocationType.PRODUCER.equals(event.getInvocationType())) {
       ProducerInvocationMonitor monitor = registryMonitor.getProducerInvocationMonitor(event.getOperationName());
       monitor.getWaitInQueue().increment(-1);
-      monitor.getLifeTimeInQueue().update(event.getInQueueNanoTime());
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java]: Why did you remove this line?

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/ConsumerInvocationMonitor.java [hunk 6]
```diff
@@ -17,11 +17,12 @@
...
 
+import java.util.HashMap;
+import java.util.Map;
 
-import org.apache.servicecomb.metrics.common.ConsumerInvocationMetric;
-import org.apache.servicecomb.metrics.common.MetricsConst;
+import org.apache.servicecomb.foundation.metrics.MetricsConst;
 
-public class ConsumerInvocationMonitor extends InvocationMonitor {
+public class ConsumerInvocationMonitor {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/ConsumerInvocationMonitor.java]: What we should do if we want to change the ConsumerInvocationMonitor into ProducerInvocationMonitor？We don't need to write lots Classes here, please use abstraction to reduce the number of the class.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/RegistryMonitor.java [hunk 7]
```diff
@@ -51,16 +48,14 @@ public ProducerInvocationMonitor getProducerInvocationMonitor(String operationNa
     return producerInvocationMonitors.computeIfAbsent(operationName, i -> new ProducerInvocationMonitor(operationName));
   }
 
-  public RegistryMetric toRegistryMetric(int windowTimeIndex) {
-    Map<String, ConsumerInvocationMetric> consumerInvocationMetrics = new HashMap<>();
+  public Map<String, Double> measure(int windowTimeIndex, boolean calculateLatency) {
+    Map<String, Double> measurements = new HashMap<>(systemMonitor.measure());
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/RegistryMonitor.java]: If we want measure the invocation monitor, we don't need to know if it consumerInvocationMonitor or producerInvocationMonitor.

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java (hunk 1 above)
Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/Invocation.java]: On producer side , startTime is when invocation received,startProcessingTime is when invocation fetch from queue and start execute;

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/DefaultHealthCheckExtraData.java (hunk 2 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/DefaultHealthCheckExtraData.java]: We may no need provider this default health check because :
1.make some  confusion
2.Service Center had same function

Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/DefaultHealthCheckExtraData.java]: Had removed

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricNode.java (hunk 3 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricNode.java]: Added

File: foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricsLoader.java (hunk 4 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/main/java/org/apache/servicecomb/foundation/metrics/publish/MetricsLoader.java]: Done

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java (hunk 5 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/event/InvocationStartProcessingEventListener.java]: queue stage latency had move from this event to finish event

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/ConsumerInvocationMonitor.java (hunk 6 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/ConsumerInvocationMonitor.java]: I had use DefaultMonitorRegistry as monitor container replace it;
In order to update monitor,I had add a group of basic wrapped monitor extends from AbstractMonitor;
Measure metrics will direct from DefaultMonitorRegistry.getInstance().getRegisteredMonitors() then getValue()

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/RegistryMonitor.java (hunk 7 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/monitor/RegistryMonitor.java]: Done

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java [hunk 8]
```diff
@@ -185,18 +185,18 @@ public String getMicroserviceQualifiedName() {
...
   }
 
-  public void triggerStartProcessingEvent() {
-    this.startProcessingTime = System.nanoTime();
-    EventUtils.triggerEvent(new InvocationStartProcessingEvent(
-        operationMeta.getMicroserviceQualifiedName(), this.invocationType));
+  public void triggerStartExecutionEvent() {
+    if (InvocationType.PRODUCER.equals(invocationType)) {
+      this.startExecutionTime = System.nanoTime();
+      EventUtils.triggerEvent(new InvocationStartExecutionEvent(operationMeta.getMicroserviceQualifiedName()));
```
Comment (WillemJiang) [inline core/src/main/java/org/apache/servicecomb/core/Invocation.java]: This EventUtils is more like a singleton.

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java [hunk 9]
```diff
@@ -87,10 +87,10 @@ public static void run() {
       Map<String, Double> metrics = restTemplate.getForObject(prefix + "/metrics", Map.class);
 
       TestMgr
-          .check(true, metrics.get("jvm(statistic=gauge,name=heapUsed)") != 0);
+          .check(true, metrics.get("jvm(name=heapUsed,statistic=gauge)") != 0);
       TestMgr.check(true, metrics.size() > 0);
       TestMgr.check(true, metrics.get(
-          "servicecomb.invocation(operation=springmvc.codeFirst.saySomething,role=producer,stage=whole,statistic=count,status=200)")
+          "servicecomb.invocation(operation=springmvc.codeFirst.saySomething,role=PRODUCER,stage=total,statistic=count,status=200)")
```
Comment (WillemJiang) [inline demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java]: the PRODUCER is upper case, but other tags are lower case.

File: demo/perf/src/main/java/org/apache/servicecomb/demo/perf/PerfMetricsFilePublisher.java [hunk 10]
```diff
@@ -75,51 +70,57 @@ private void collectVertxMetrics(MetricsLoader loader, StringBuilder sb) {
...
-            MetricsConst.TAG_STATUS);
+    MetricNode treeNode;
+    try {
+      treeNode = loader
+          .getMetricTree(MetricsConst.SERVICECOMB_INVOCATION, MetricsConst.TAG_ROLE, MetricsConst.TAG_OPERATION,
+              MetricsConst.TAG_STATUS);
+    }
+    //before receive any request,there are no MetricsConst.SERVICECOMB_INVOCATION,so getMetricTree will throw ServiceCombException
+    catch (ServiceCombException ignored) {
+      return;
```
Comment (WillemJiang) [inline demo/perf/src/main/java/org/apache/servicecomb/demo/perf/PerfMetricsFilePublisher.java]: We need to log the exception for trace the issue

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestMetricNode {
+  @Test
+  public void test() {
```
Comment (WillemJiang) [inline foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java]: The unit test method is not generic one, it should has some meaning which can help us to find out the key reason of failed test.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    setupWindowTime();
+    registerSystemMetrics();
+  }
+
+  private void setupWindowTime() {
+    int time = DynamicPropertyFactory.getInstance().getIntProperty(MetricsConfig.METRICS_WINDOW_TIME, 5000).get();
+    System.getProperties().setProperty("servo.pollers", time > 0 ? String.valueOf(time) : "5000");
+  }
+
+  public Counter getCounter(boolean isStepCounter, String name, String... tags) {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java]: This method exposes the counter detail which is not right way.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java [hunk 13]
```diff
@@ -60,16 +53,19 @@ private void init(HealthCheckerManager manager) {
...
 
-  @ApiResponses({
-      @ApiResponse(code = 400, response = String.class, message = "illegal request content"),
-  })
-  @RequestMapping(path = "/{name}", method = RequestMethod.GET)
+  @RequestMapping(path = "/detail", method = RequestMethod.GET)
   @CrossOrigin
-  public HealthCheckResult healthWithName(@PathVariable(name = "name") String name) {
-    return manager.check(name);
+  public Map<String, HealthCheckResult> checkHealthDetail() {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java]: Details

File: core/src/main/java/org/apache/servicecomb/core/Invocation.java (hunk 8 above)
Comment (zhengyangyong) [reply core/src/main/java/org/apache/servicecomb/core/Invocation.java]: OK , delete EventUtils and make EventBus as a singleton

File: demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java (hunk 9 above)
Comment (zhengyangyong) [reply demo/demo-springmvc/springmvc-client/src/main/java/org/apache/servicecomb/demo/springmvc/client/SpringmvcClient.java]: Done,add toLowerCase()

File: demo/perf/src/main/java/org/apache/servicecomb/demo/perf/PerfMetricsFilePublisher.java (hunk 10 above)
Comment (zhengyangyong) [reply demo/perf/src/main/java/org/apache/servicecomb/demo/perf/PerfMetricsFilePublisher.java]: add a containsId method and no need throw exception any more

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java (hunk 11 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java]: Ok ,add comments for explain check reason

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java (hunk 12 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/MonitorManager.java]: separate into two method named getCounter and getStepCounter

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java (hunk 13 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java]: Fixed

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java (hunk 11 above)
Comment (WillemJiang) [reply foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java]: You need to separate the test methods, please take https://github.com/apache/incubator-servicecomb-saga/blob/master/alpha/alpha-core/src/test/java/org/apache/servicecomb/saga/alpha/core/CompositeOmegaCallbackTest.java as an example.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java [hunk 14]
```diff
@@ -66,6 +47,6 @@ public boolean checkHealth() {
   @RequestMapping(path = "/detail", method = RequestMethod.GET)
```
Comment (WillemJiang) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java]: The path name need to updated.

File: metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java [hunk 15]
```diff
@@ -30,8 +30,7 @@
 
   @Test
   public void test() {
-    HealthCheckerManager manager = new HealthCheckerManager();
-    manager.register(new HealthChecker() {
+    HealthCheckerManager.getInstance().register(new HealthChecker() {
```
Comment (WillemJiang) [inline metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java]: Here you just check one HealthChecker. The test should test at least zero，one and two HealthChecker.

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java (hunk 14 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java]: change to details

File: metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java (hunk 15 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java]: Done

File: foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java (hunk 11 above)
Comment (zhengyangyong) [reply foundations/foundation-metrics/src/test/java/org/apache/servicecomb/foundation/metrics/publish/TestMetricNode.java]: OK

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java (hunk 14 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java]: finnal path is /health/details

File: metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java (hunk 15 above)
Comment (zhengyangyong) [reply metrics/metrics-core/src/test/java/org/apache/servicecomb/metrics/core/TestHealthCheckerPublisher.java]: zero，one and two HealthChecker check logic had added in TestHealthCheckerManager

File: metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.servicecomb.provider.rest.common.RestSchema;
+import org.springframework.web.bind.annotation.CrossOrigin;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+@RestSchema(schemaId = "healthEndpoint")
+@RequestMapping(path = "/health")
+public class HealthCheckerPublisher {
+  @RequestMapping(path = "/", method = RequestMethod.GET)
+  @CrossOrigin
```
Comment (liubao68) [inline metrics/metrics-core/src/main/java/org/apache/servicecomb/metrics/core/publish/HealthCheckerPublisher.java]: This annotation may not work
